cluster/tls_transport.go

```diff
@@ -0,0 +1,341 @@
+	if ip != "" {
+		advertiseAddr = net.ParseIP(ip)
+		if advertiseAddr == nil {
+			return nil, 0, fmt.Errorf("failed to parse advertise address %q", ip)
+		}
+
+		if ip4 := advertiseAddr.To4(); ip4 != nil {
+			advertiseAddr = ip4
+		}
+		advertisePort = port
+	} else {
+		if t.bindAddr == "0.0.0.0" {
+			// Otherwise, if we're not bound to a specific IP, let's
+			// use a suitable private IP address.
+			var err error
+			ip, err = sockaddr.GetPrivateIP()
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to get interface addresses: %v", err)
+			}
+			if ip == "" {
+				return nil, 0, fmt.Errorf("no private IP address found, and explicit IP not provided")
+			}
+
+			advertiseAddr = net.ParseIP(ip)
+			if advertiseAddr == nil {
+				return nil, 0, fmt.Errorf("failed to parse advertise address: %q", ip)
+			}
+		} else {
+			advertiseAddr = t.listener.Addr().(*net.TCPAddr).IP
+		}
+		advertisePort = t.GetAutoBindPort()
+	}
+	return advertiseAddr, advertisePort, nil
+}
+
+// PacketCh returns a channel that can be read to receive incoming
+// packets from other peers.
+func (t *TLSTransport) PacketCh() <-chan *memberlist.Packet {
+	return t.packetCh
+}
+
+// StreamCh returns a channel that can be read to handle incoming stream
+// connections from other peers.
+func (t *TLSTransport) StreamCh() <-chan net.Conn {
+	return t.streamCh
+}
+
+// Shutdown is called when memberlist is shutting down; this gives the
+// TLS Transport a chance to clean up the listener and other goroutines.
+func (t *TLSTransport) Shutdown() error {
+	level.Debug(t.logger).Log("msg", "shutting down tls transport")
+	t.cancel()
+	err := t.listener.Close()
+	t.connPool.shutdown()
+	<-t.done
+	return err
+}
+
+// WriteTo is a packet-oriented interface that borrows a connection
+// from the pool, and writes to it. It also returns a timestamp of when
+// the packet was written.
+func (t *TLSTransport) WriteTo(b []byte, addr string) (time.Time, error) {
+	conn, err := t.connPool.borrowConnection(addr, DefaultTcpTimeout)
+	if err != nil {
+		t.writeErrs.WithLabelValues("packet").Inc()
+		return time.Now(), errors.Wrap(err, "failed to dial")
+	}
+	fromAddr := t.listener.Addr().String()
+	err = conn.writePacket(fromAddr, b)
+	if err != nil {
+		t.writeErrs.WithLabelValues("packet").Inc()
+		return time.Now(), errors.Wrap(err, "failed to write packet")
+	}
+	t.packetsSent.Add(float64(len(b)))
+	return time.Now(), nil
+}
+
+// DialTimeout is used to create a connection that allows memberlist
+// to perform two-way communications with a peer.
+func (t *TLSTransport) DialTimeout(addr string, timeout time.Duration) (net.Conn, error) {
+	conn, err := dialTLSConn(addr, timeout, t.tlsClientCfg)
+	if err != nil {
+		t.writeErrs.WithLabelValues("stream").Inc()
+		return nil, errors.Wrap(err, "failed to dial")
+	}
+	err = conn.writeStream()
+	if err != nil {
+		t.writeErrs.WithLabelValues("stream").Inc()
+		return conn.connection, errors.Wrap(err, "failed to create stream connection")
```

Comment:
Similar thoughts here - perhaps, we're assuming we want to sync writing but not reading - is this correct?

Reply:
I think this is a similar situation to the above. The connection is never reused so it is incidentally working. I'm adding a new function and some safety measures.